zkevm-circuits/src/evm_circuit/op_execution/context.rs

```diff
@@ -0,0 +1,284 @@
+use super::super::{BusMappingLookup, Cell, Constraint, FixedLookup, Lookup};
+use super::OpExecutionState;
+use crate::util::Expr;
+use halo2::{arithmetic::FieldExt, plonk::Expression};
+
+pub const STACK_START_IDX: usize = 1024;
+
+// Allows updating the state transition without having to update all
+// op codes that don't modify any of the new state variables.
+#[derive(Clone, Debug, Default)]
+pub struct StateTransitions<F> {
+    pub gc_delta: Option<Expression<F>>,
+    pub sp_delta: Option<Expression<F>>,
+    pub pc_delta: Option<Expression<F>>,
+    pub gas_delta: Option<Expression<F>>,
+    pub next_memory_size: Option<Expression<F>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Context<F> {
+    pub expressions: Vec<Expression<F>>,
+    pub(super) lookups: Vec<Lookup<F>>,
+    pub call_id: Expression<F>,
+    pub stack_offset: i32,
+}
+
+impl<F: FieldExt> Context<F> {
```

Comment:
does it stand for callContext ?

Reply:
Sorry for the bad name. The idea is to have something like `ConstraintSystem` or `ProtoBoard` which we can more easily use in the op code gadgets to generate the necessary constraints. This should also help for code reuse/duplication and readability. Currently named it `ConstraintBuilder` which is hopefully a better name. 

Easy example: instead of doing this in all the gadgets to pop a value of the stack

```
Lookup::BusMappingLookup(BusMappingLookup::Stack {
    index_offset: 0,
    value: value.expr(),
    is_write: false,
})
```
we can now just write

```
cb.stack_pop(value.expr());
```

Easier to read, no way to forget to include the constraint that is being built and if something changes to the busmapping API there's only a single place where the code needs to be updated. So just a simple extra layer of abstraction which I think is useful to do for a couple of reasons.